### PR TITLE
feat: allow merging PRs with Github's auto-merge

### PR DESCRIPTION
**Pull request by jashandeep-sohi**
This adds support for Github's auto-merge feature:

- https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-auto-merge-for-pull-requests-in-your-repository
- https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/automatically-merging-a-pull-request

This can only be done via the GraphQL API, so I've had to vendor another library:

https://docs.github.com/en/graphql/reference/mutations#enablepullrequestautomerge

To maintain backwards compatibilty, the exisiting method of merging a PR is left intact and as the default.

Two new flags are introduced, `--pr-merge-auto` & `--pr-merge-auto-wait`.

`--pr-merge-auto` enables "auto-merge" for the PR (via a GraphQL mutation). This does not wait for the PR to actually be merged. That's left up to Github once all branch protection rules pass.

`--pr-merge-auto-wait` can optionally wait until the PR is actually merged by Github.

Both of these flags are also exposed on a per repository basis using the `mergeauto` & `mergeautowait` options.

# Notes
- PRs must be in a non "clean" state when auto-merge is enabled.
- Practically, this means there must be at least 1 branch protection rule that blocks the merge initially.
- Cannot enable for PRs in "draft" state
- Snuck in a `flake.nix`. Makes it easy to install/lock to any commit with Nix. e.g. `nix run github:jashandeep-sohi/octopilot/feat/github-auto-merge`. Happy to remove it if you like.
- When `--pr-merge-auto-wait` is enabled, and we timeout waiting, should auto-merge for the PR be disabled? The merge could still happen at a later point if we leave it enabled. Which might be surprising.

### CODE REVIEW

**Comment by vbehar** [comment]
> When --pr-merge-auto-wait is enabled, and we timeout waiting, should auto-merge for the PR be disabled? The merge could still happen at a later point if we leave it enabled. Which might be surprising.

Yes I think we should avoid this kind of surprise, and disable auto-merge if we timeout.
maybe later we can add another option to leave it as-is - but if people want to use an async auto merge, they can just set the merge-auto-wait flag to false

File: repository/pull_request.go [hunk 1]
```diff
@@ -275,27 +277,185 @@ func (r Repository) addPullRequestAssignees(ctx context.Context, options GitHubO
...
+		PullRequestID:   pr.NodeID,
+		MergeMethod:     &mergeMethod,
+		ExpectedHeadOid: expectedHeadOid,
+		CommitHeadline:  commitHeadLine,
+		CommitBody:      commitBody,
+	}
+
+	attempted := 0
+
+	// Should this be retried?
```
Comment (vbehar) [inline repository/pull_request.go:386]: I'm ok with retrying

File: repository/pull_request.go [hunk 2]
```diff
@@ -275,27 +277,185 @@ func (r Repository) addPullRequestAssignees(ctx context.Context, options GitHubO
...
+		}
+
+		logrus.WithFields(logrus.Fields{
+			"repository":     r.FullName(),
+			"pull-request":   prURL,
+			"attempted":      attempted,
+			"merge-strategy": mergeStrategy,
+		}).WithError(err).Warning("Failed to enable auto-merge for Pull Request - will retry")
+
+		// To wait or not to wait?
```
Comment (vbehar) [inline repository/pull_request.go:407]: same as retrying, it's ok for me

**Review by vbehar — APPROVED** [review]

**Comment by vbehar** [comment]
thanks for your contribution!

I'll just let you remove the auto-merge if we time-out waiting, otherwise it's good for me

**Comment by MPV** [comment]
Just doublechecking: Does this require the auto merge feature to be enabled on the repos getting the PR? If yes, does it also set it? I have thousands of repos where I'd like to automerge, but their configs may differ. Curious about if/how you've solved this?

**Comment by jashandeep-sohi** [comment]
> Just doublechecking: Does this require the auto merge feature to be enabled on the repos getting the PR? If yes, does it also set it? I have thousands of repos where I'd like to automerge, but their configs may differ. Curious about if/how you've solved this?

Yes, it requires the auto merge feature to be enabled on the repo.
This is a one-time step, but you bring up a good point about updating thousands of repos.

If you're managing your repo settings with terraform or the like, maybe you could do it there.
The `gh` CLI does allow you to do this one a per repo basis, so you could also do something like:
`gh search repos --owner=jashandeep-sohi "octopilot" --json url --jq '.[].url' | xargs -L1 gh repo edit --enable-auto-merge`

I think we could attempt to do it within octopilot every time, if not enabled.
I believe this can be done via the REST API by PATCHing [`allow_auto_merge`](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#update-a-repository). But this means needing **write** [repo admin permissions](https://docs.github.com/en/rest/overview/permissions-required-for-github-apps?apiVersion=2022-11-28#repository-permissions-for-administration), which I was trying to avoid.

**Review by vbehar — APPROVED** [review]
thanks, sounds good to me

**Pull request merged**